### ISSUE TITLE
Evaluate Profile-Guided Optimization (PGO)

### ISSUE THREAD

**Issue by zamazan4ik** [containers/youki#2386]
Hi!

Here I am posting an idea for optimizing Youki with Profile-Guided Optimization (PGO). Recently I started evaluating PGO across multiple software domains - all my current results are available here: https://github.com/zamazan4ik/awesome-pgo . For Youki I did some quick benchmarks on my local Linux machine and want to share the actual performance numbers.

## Test environment

* Fedora 38
* Linux kernel 6.4.15-200.fc38.x86_64
* AMD Ryzen 9 5900x
* 48 Gib RAM
* SSD Samsung 980 Pro 2 Tib
* Rust `rustc 1.72.0 (5680fa18f 2023-08-23)`
* Youki version: latest commit (`646c1034f78454904cc3e1ccec2cd8dc270ab3fd` commit) in the `main` branch

## Benchmark

As a benchmark, I use the suggested in the README file workload with `sudo ./youki create -b tutorial a && sudo ./youki start a && sudo ./youki delete -f a`

`youki_release` is built with `just youki-release`. PGO optimized build is done with [cargo-pgo](https://github.com/Kobzol/cargo-pgo) (`cargo pgo build` + run the benchmark with the Instrumented Youki + `cargo pgo optimize build`). As a training workload, I use the benchmark itself. 

## Results

The results are presented in `hyperfine` format. All benchmarks are done multiple times, in different order, etc - the results are reproducible.

```
sudo hyperfine --prepare 'sudo sync; echo 3 | sudo tee /proc/sys/vm/drop_caches' --warmup 100 --min-runs 500 'sudo ./youki_release create -b tutorial a && sudo ./youki_release start a && sudo ./youki_release delete -f a' 'sudo ./youki_optimized create -b tutorial a && sudo ./youki_optimized start a && sudo ./youki_optimized delete -f a'
Benchmark 1: sudo ./youki_release create -b tutorial a && sudo ./youki_release start a && sudo ./youki_release delete -f a
  Time (mean ± σ):      78.6 ms ±   3.7 ms    [User: 11.2 ms, System: 43.9 ms]
  Range (min … max):    70.9 ms …  97.8 ms    500 runs

Benchmark 2: sudo ./youki_optimized create -b tutorial a && sudo ./youki_optimized start a && sudo ./youki_optimized delete -f a
  Time (mean ± σ):      77.4 ms ±   3.6 ms    [User: 10.9 ms, System: 44.1 ms]
  Range (min … max):    70.6 ms …  90.0 ms    500 runs

Summary
  sudo ./youki_optimized create -b tutorial a && sudo ./youki_optimized start a && sudo ./youki_optimized delete -f a ran
    1.02 ± 0.07 times faster than sudo ./youki_release create -b tutorial a && sudo ./youki_release start a && sudo ./youki_release delete -f a
```
Just for reference, I also share the results for Instrumentation mode:
```
LLVM_PROFILE_FILE=/home/zamazan4ik/open_source/youki/target/pgo-profiles/youki_%m_%p.profraw sudo hyperfine --prepare 'sudo sync; echo 3 | sudo tee /proc/sys/vm/drop_caches' --warmup 10 --min-runs 100 'sudo ./youki_instrumented create -b tutorial a && sudo ./youki_instrumented start a && sudo ./youki_instrumented delete -f a'
Benchmark 1: sudo ./youki_instrumented create -b tutorial a && sudo ./youki_instrumented start a && sudo ./youki_instrumented delete -f a
  Time (mean ± σ):     161.1 ms ±   3.3 ms    [User: 20.3 ms, System: 116.8 ms]
  Range (min … max):   154.8 ms … 170.7 ms    100 runs
```

According to the tests, PGO helps with achieving quite better performance (1-2%). Not a great win but it's not bad "just" for a compiler option. On a scale, even 1% is a good thing to achieve.

## Further steps
If you think that it's worth it, I think we can perform more robust PGO benchmarks for Youki. And then document the results of the project. So other people will be able to optimize Youki for their own workloads.

**Comment by utam0k** [comment]
@zamazan4ik Could you write the documentation on how to optimize with PGO in our official documentation?
https://containers.github.io/youki/youki.html

**Comment by zamazan4ik** [comment]
I think I can. But do you want to publish this documentation based only on the tests above? Do you think that PGO according to the results above worth it?

**Comment by utam0k** [comment]
@zamazan4ik 
To be honest, I am too much of a beginner to know how much performance improvement I should expect. Anyway, I'd like to see the result of this.

> If you think that it's worth it, I think we can perform more robust PGO benchmarks for Youki.

**Comment by zamazan4ik** [comment]
> To be honest, I am too much of a beginner to know how much performance improvement I should expect. Anyway, I'd like to see the result of this.

I already posted some results in the starting post with my current results:
```
sudo hyperfine --prepare 'sudo sync; echo 3 | sudo tee /proc/sys/vm/drop_caches' --warmup 100 --min-runs 500 'sudo ./youki_release create -b tutorial a && sudo ./youki_release start a && sudo ./youki_release delete -f a' 'sudo ./youki_optimized create -b tutorial a && sudo ./youki_optimized start a && sudo ./youki_optimized delete -f a'
Benchmark 1: sudo ./youki_release create -b tutorial a && sudo ./youki_release start a && sudo ./youki_release delete -f a
  Time (mean ± σ):      78.6 ms ±   3.7 ms    [User: 11.2 ms, System: 43.9 ms]
  Range (min … max):    70.9 ms …  97.8 ms    500 runs

Benchmark 2: sudo ./youki_optimized create -b tutorial a && sudo ./youki_optimized start a && sudo ./youki_optimized delete -f a
  Time (mean ± σ):      77.4 ms ±   3.6 ms    [User: 10.9 ms, System: 44.1 ms]
  Range (min … max):    70.6 ms …  90.0 ms    500 runs

Summary
  sudo ./youki_optimized create -b tutorial a && sudo ./youki_optimized start a && sudo ./youki_optimized delete -f a ran
    1.02 ± 0.07 times faster than sudo ./youki_release create -b tutorial a && sudo ./youki_release start a && sudo ./youki_release delete -f a
```

`youki_release` - the default Release build, `youki_optimized` - Release + PGO optimization version.

**Comment by zamazan4ik** [comment]
> Thanks for your guide. 1ms probably doesn't have a strong meaning in the real world.

it's up to you :) However, if Youki really cares about performance, it's not a bad thing to automatically increase the performance for 1-2% just with a compiler option.

> How many times can you run it and still arrive at the same result?

As you see, `hyperfine` already runs the test workload multiple times (500 runs + 100 for warmup). Regarding the whole experiment - I did it multiple times (more than 3 times), in different order between binaries - the results are the same.

**Comment by utam0k** [comment]
> As you see, hyperfine already runs the test workload multiple times (500 runs + 100 for warmup). Regarding the whole experiment - I did it multiple times (more than 3 times), in different order between binaries - the results are the same.

👍 

> it's up to you :) However, if Youki really cares about performance, it's not a bad thing to automatically increase the performance for 1-2% just with a compiler option.

I'm curious about multiple kernel versions. Is this compilation kernel-independent? It would be attractive if it did not.

**Comment by zamazan4ik** [comment]
> I'm curious about multiple kernel versions. Is this compilation kernel-independent? It would be attractive if it did not.

Yes, it does not depend on the kernel. So it would be nice if you can reproduce the results on different setups.
